pkg/shed/db.go

```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+// given function to see if it needs to stop the iteration or not. The skipStartKey indicates
+// weather to skip the first key or not.
+func (db *DB) Iterate(startKey []byte, skipStartKey bool, fn func(key []byte, value []byte) (stop bool, err error)) (err error) {
+	if startKey != nil {
+		db.logger.Tracef("Iterating with startKey %s and skipping key is %t", string(startKey), skipStartKey)
+	}
+
+	db.metrics.IterationCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = true
+		o.PrefetchSize = 1024
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		i.Seek(startKey)
+		if !i.Valid() {
+			return nil
+		}
+
+		if skipStartKey {
+			i.Next()
+		}
+
+		for ; i.Valid(); i.Next() {
+			item := i.Item()
+			k := item.Key()
+			v, err := item.ValueCopy(nil)
+			if err != nil {
+
+				return err
+			}
+			stop, err := fn(k, v)
+			if err != nil {
+				return err
+			}
+			if stop {
+				return nil
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while doing iteration. startKey : %v, skipStartKey: %t",
+			string(startKey), skipStartKey)
+		db.logger.Debugf("error while doing iteration. startKey : %v, skipStartKey: %t, Error : %w",
+			string(startKey), skipStartKey, err.Error())
+		db.metrics.IterationFailCount.Inc()
+	}
+	return err
+}
+
+// First returns the first key which matches the given prefix.
+func (db *DB) First(prefix []byte) (key []byte, value []byte, err error) {
+	if prefix != nil {
+		db.logger.Tracef("Finding first key with prefix %s", string(prefix))
+	}
+	db.metrics.FirstCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = true
+		o.PrefetchSize = 1
+
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		i.Seek(prefix)
+		key = i.Item().Key()
+		if !bytes.HasPrefix(key, prefix) {
+			return ErrNotFound
+		}
+		value, err = i.Item().ValueCopy(value)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while finding first key with prefix : %v, Error : %w", string(prefix), err.Error())
+		db.logger.Debugf("error while doing first. prefix : %v, Error : %w", string(prefix), err.Error())
+		db.metrics.FirstFailCount.Inc()
+	} else {
+		db.logger.Tracef("first value with prefix %s, key %s, value len %d", string(prefix), string(key), len(value))
+	}
+
+	return key, value, err
+}
+
+// Last retuns the last key matching the given prefix.
+func (db *DB) Last(prefix []byte) (key []byte, value []byte, err error) {
+	if prefix != nil {
+		db.logger.Tracef("Finding last key with prefix %s", string(prefix))
+	}
+	db.metrics.LastCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = true
+		o.PrefetchSize = 1024
+
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		for i.Seek(prefix); i.ValidForPrefix(prefix); i.Next() {
```

Comment:
This will iterate over all keys with prefixes to get the last one. Looks expensive.

Shed in swarm uses more efficient approach to seek ahead and have only one Prev call afterwards https://github.com/ethersphere/swarm/blob/master/shed/index.go#L375. Why was that abandoned?

Reply:
i also thought about this.. There is no prev() call in badger. i can raise a request for them. until then this is the only way.

Reply:
Couldn't the `IteratorOptions.Reverse` option be used to get the Prev functionality?

Reply:
did that exactly